lib/core/test/unit/Cardano/WalletSpec.hs

```diff
@@ -627,6 +687,321 @@ prop_estimateFee (NonEmpty coins) =
+    :: GenTxHistory
+    -> [LocalTxSubmissionStatus SealedTx]
+mkLocalTxSubmissionStatus = mapMaybe getStatus . getTxHistory
+  where
+    getStatus :: (Tx, TxMeta) -> Maybe (LocalTxSubmissionStatus SealedTx)
+    getStatus (tx, txMeta)
+        | isPending txMeta = Just st
+        | otherwise = Nothing
+      where
+        i = tx ^. #txId
+        sl = txMeta ^. #slotNo
+        st = LocalTxSubmissionStatus i (SealedTx (getHash i)) sl sl
+
+instance Arbitrary SlottingParameters where
+    arbitrary = mk <$> choose (0.5, 1)
+      where
+        mk f = dummySlottingParameters
+            { getActiveSlotCoefficient = ActiveSlotCoefficient f }
+
+-- | 'WalletLayer' context.
+data TxRetryTestCtx = TxRetryTestCtx
+    { ctxDbLayer :: DBLayer TxRetryTestM DummyState ShelleyKey
+    , ctxNetworkLayer :: NetworkLayer TxRetryTestM Block
+    , ctxTracer :: Tracer IO W.WalletLog
+    , ctxWalletId :: WalletId
+    } deriving (Generic)
+
+data TxRetryTestState = TxRetryTestState
+    { testCase :: TxRetryTest
+    , timeVar :: MVar Time
+    , timeStep :: DiffTime
+    } deriving (Generic)
+
+data TxRetryTestResult a = TxRetryTestResult
+    { resLogs :: [W.WalletLog]
+    , resAction :: a
+    , resSubmittedTxs :: [SealedTx]
+    } deriving (Generic, Show, Eq)
+
+newtype TxRetryTestM a = TxRetryTestM
+    { unTxRetryTestM :: ReaderT TxRetryTestState IO a
+    } deriving (Functor, Applicative, Monad, MonadIO, MonadFail)
+
+instance MonadUnliftIO TxRetryTestM where
+    withRunInIO = wrappedWithRunInIO TxRetryTestM unTxRetryTestM
+
+instance MonadMonotonicTime TxRetryTestM where
+    getMonotonicTime = do
+        st <- TxRetryTestM ask
+        modifyMVar (timeVar st) $ \t -> do
+            let t' = addTime (timeStep st) t
+            pure (t', t')
+
+instance MonadTime TxRetryTestM where
+    getCurrentTime = liftIO getCurrentTime
+
+prop_localTxSubmission :: TxRetryTest -> Property
+prop_localTxSubmission tc = monadicIO $ do
+    timeVar <- newMVar (Time 0)
+    let st = TxRetryTestState tc timeVar 2
+    res <- run $ runTest st $ \ctx@(TxRetryTestCtx DBLayer{..} _ _ wid) -> do
+        atomically $ do
+            let txHistory = getTxHistory (retryTestTxHistory tc)
+            unsafeRunExceptT $ putTxHistory (PrimaryKey wid) txHistory
+            forM_ (retryTestPool tc) $ \(LocalTxSubmissionStatus txid tx _ sl) ->
+                unsafeRunExceptT $ putLocalTxSubmission (PrimaryKey wid) txid tx sl
+
+        runLocalTxSubmissionPool' @_ @DummyState @ShelleyKey (timeStep st) 10 ctx wid
+        atomically $ readLocalTxSubmissionPending (PrimaryKey wid)
+
+    monitor $ counterexample $ unlines $
+        [ "posted txs = " ++ show (resSubmittedTxs res)
+        , "final pool state = " ++ show (resAction res)
+        , "logs:"
+        ] ++ map (T.unpack . toText) (resLogs res)
+
+    -- props:
+    --  1. pending transactions in pool are retried
+    assert (all (`elem` (submittedTx <$> retryTestPool tc)) (resSubmittedTxs res))
```

Comment:
:+1:

Reply:
This turned out to be a lot of effort to set up the test, for some fairly weak properties. :grimacing: 
But I'm reasonably satisfied that it works.

Reply:
I have to say it was also not easy to review 😅, took me some time to get the gist but I ended up being convinced by the approach.